lib/semantics/tools.cc

```diff
@@ -443,6 +490,45 @@ bool IsAssumedSizeArray(const Symbol &symbol) {
   return details && details->IsAssumedSize();
 }
 
+bool IsExternalInPureContext(const Symbol &symbol, const Scope &scope) {
+  if (const auto *pureProc{semantics::FindPureProcedureContaining(&scope)}) {
+    if (const Symbol * root{GetAssociationRoot(symbol)}) {
+      if (semantics::FindExternallyVisibleObject(*root, *pureProc)) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+bool InProtectedContext(const Symbol &symbol, const Scope &currentScope) {
+  return IsProtected(symbol) && !IsHostAssociated(symbol, currentScope);
+}
+
+std::optional<parser::MessageFixedText> WhyNotModifiable(
+    const Symbol &symbol, const Scope &scope) {
+  const Symbol *root{GetAssociationRoot(symbol)};
+  if (!root) {
+    return "'%s' is construct associated with an expression"_en_US;
+  } else if (InProtectedContext(*root, scope)) {
+    return "'%s' is protected in this scope"_en_US;
+  } else if (IsExternalInPureContext(*root, scope)) {
+    return "'%s' is an external that is referenced in a PURE procedure"_en_US;
+  } else if (IsOrContainsEventOrLockComponent(*root)) {
+    return "'%s' is an entity with either an EVENT_TYPE or LOCK_TYPE"_en_US;
+  } else if (IsIntentIn(*root) && !IsValueDummy(*root)) {
+    return "'%s' is an INTENT(IN) dummy argument"_en_US;
+  } else if (!IsVariableName(*root)) {
+    return "'%s' is not a variable"_en_US;
+  } else {
+    return std::nullopt;
+  }
+}
+
+bool IsModifiable(const Symbol &symbol, const Scope &scope) {
```

Comment:
Is this ever used? It seems like we'll always want the reason.

Reply:
I used it in earlier code, but since it's not currently used, I'll remove it.